requirements.txt

```diff
@@ -0,0 +1,4 @@
+ops >= 1.2.0
```

Comment:
This was intentional as it required at least ops 1.2.0.  Of course `>=` is more fragile than `==`... I'm constantly torn between predictability (completely pinned) and reducing maintenance efforts (eg: us having to bump every charm we had to `ops==1.2.0` manually because we wanted to fix the relation data bug).  

I lean toward `>=X` unless there's a reason to pin, provided we do nightly or weekly CI builds.  Things may break, but we can see when and figure it out.  Thoughts?

Reply:
I don't think we need to go completely pinned, something like `<1.3.0` would catch e.g. `1.2.1` getting published. The minor version breaking things happens fairly often though, from Kubernetes itself (1.22 broke lots of things) to charmcraft (1.3 refuses to build without a `charmcraft.yaml` file). It breaks the semantic version ideal as defined by semver.org, but lots of projects like to introduce breaking changes in 1.X versions.

Maybe using `<` and something like dependabot is the sweet spot?

Reply:
I think this should be `ops<1.3.0`

Reply:
I've accepted this suggestion, but I think using this approach bakes in maintenance for us.  To me the better approach is loose ranges (eg: only use `<`/`!=`/whatever when you know there's something to avoid) with good CI coverage and weekly testing.  If an update breaks CI, then we know and can fix it

Even better is if we have something to make the update PRs for us though (renovate, or similar).